src/Microsoft.Health.Dicom.Core/Features/Store/StoreOrchestrator.cs

```diff
@@ -115,4 +144,92 @@ private async Task TryCleanupInstanceIndexAsync(VersionedInstanceIdentifier vers
             // Fire and forget.
         }
     }
+
+    internal static Dictionary<int, FrameRange> GetFramesOffset(DicomDataset dataset)
+    {
+        if (!dataset.TryGetPixelData(out DicomPixelData dicomPixel))
+        {
+            return null;
+        }
+
+        if (dicomPixel.NumberOfFrames < 1)
+        {
+            return null;
+        }
+
+        var pixelData = dataset.GetDicomItem<DicomItem>(DicomTag.PixelData);
+        var framesRange = new Dictionary<int, FrameRange>();
+
+        // todo support case where fragments != frames.
+        // This means offsettable matches the frames and we have to parse the bytes in pixelData to find the right fragment and end at the right fragment.
+        // there is also a 8 byte tag inbetween the fragment data that we need to handlee.
+        // fo-dicom/DicomPixelData.cs/GetFrame has the logic
+        if (pixelData is DicomFragmentSequence pixelDataFragment
+            && pixelDataFragment.Fragments.Count == dicomPixel.NumberOfFrames)
+        {
+            for (int i = 0; i < pixelDataFragment.Fragments.Count; i++)
+            {
+                var fragment = pixelDataFragment.Fragments[i];
+                if (TryGetBufferPosition(fragment, out long position, out long size))
+                {
+                    framesRange.Add(i, new FrameRange(position, size));
+                }
+            }
+        }
+        else if (pixelData is DicomOtherByte)
+        {
+            var dicomPixelOtherByte = dataset.GetDicomItem<DicomOtherByte>(DicomTag.PixelData);
+
+            for (int i = 0; i < dicomPixel.NumberOfFrames; i++)
+            {
+                IByteBuffer byteBuffer = dicomPixel.GetFrame(i);
+                if (TryGetBufferPosition(dicomPixelOtherByte.Buffer, out long position, out long size)
+                    && byteBuffer is RangeByteBuffer rangeByteBuffer)
+                {
+                    framesRange.Add(i, new FrameRange(position + rangeByteBuffer.Offset, rangeByteBuffer.Length));
+                }
+            }
+        }
+        else if (pixelData is DicomOtherWord)
+        {
+            var dicomPixelWordByte = dataset.GetDicomItem<DicomOtherWord>(DicomTag.PixelData);
+
+            for (int i = 0; i < dicomPixel.NumberOfFrames; i++)
+            {
+                IByteBuffer byteBuffer = dicomPixel.GetFrame(i);
+                if (TryGetBufferPosition(dicomPixelWordByte.Buffer, out long position, out long size)
+                    && byteBuffer is RangeByteBuffer rangeByteBuffer)
+                {
+                    framesRange.Add(i, new FrameRange(position + rangeByteBuffer.Offset, rangeByteBuffer.Length));
+                }
+            }
+        }
+
+        if (framesRange.Any())
+        {
+            return framesRange;
+        }
+
+        return null;
+    }
+
+    private static bool TryGetBufferPosition(IByteBuffer buffer, out long position, out long size)
+    {
+        bool result = false;
```

Comment:
nit: can do these default assignments in a final `else` block

Reply:
General guidlines is to set the out parameters default first.